spark/src/main/java/org/apache/iceberg/spark/arrow/ArrowUtils.java

```diff
@@ -0,0 +1,113 @@
+
+import org.apache.arrow.memory.RootAllocator;
+import org.apache.arrow.vector.types.DateUnit;
+import org.apache.arrow.vector.types.FloatingPointPrecision;
+import org.apache.arrow.vector.types.TimeUnit;
+import org.apache.arrow.vector.types.pojo.ArrowType;
+import org.apache.arrow.vector.types.pojo.Field;
+import org.apache.spark.sql.types.ArrayType;
+import org.apache.spark.sql.types.DataType;
+import org.apache.spark.sql.types.DataTypes;
+import org.apache.spark.sql.types.DecimalType;
+import org.apache.spark.sql.types.Metadata;
+import org.apache.spark.sql.types.StructField;
+import org.apache.spark.sql.types.StructType;
+
+public class ArrowUtils {
+
+  private static ArrowUtils instance;
+  private RootAllocator rootAllocator;
+
+  private ArrowUtils() {
+    rootAllocator = new RootAllocator(Long.MAX_VALUE);
+  }
+
+  public static ArrowUtils instance() {
+    if (instance == null) {
+      instance = new ArrowUtils();
+    }
+    return instance;
+  }
+
+  public RootAllocator rootAllocator() {
+    return rootAllocator;
+  }
+
+  @SuppressWarnings("checkstyle:CyclomaticComplexity")
+  public DataType fromArrowType(ArrowType data) {
+
+    if (data instanceof ArrowType.Bool) {
+      return DataTypes.BooleanType;
+    } else if (data instanceof ArrowType.Int) {
+      ArrowType.Int intData = (ArrowType.Int) data;
+      if (intData.getIsSigned() && intData.getBitWidth() == 8) {
+        return DataTypes.ByteType;
+      } else if (intData.getIsSigned() && intData.getBitWidth() == 8 * 2) {
+        return DataTypes.ShortType;
+      } else if (intData.getIsSigned() && intData.getBitWidth() == 8 * 4) {
+        return DataTypes.IntegerType;
+      } else if (intData.getIsSigned() && intData.getBitWidth() == 8 * 8) {
+        return DataTypes.LongType;
+      }
+    } else if (data instanceof ArrowType.FloatingPoint) {
+      ArrowType.FloatingPoint floatData = (ArrowType.FloatingPoint) data;
+      if (floatData.getPrecision() == FloatingPointPrecision.SINGLE) {
+        return DataTypes.FloatType;
+      } else if (floatData.getPrecision() == FloatingPointPrecision.DOUBLE) {
+        return DataTypes.DoubleType;
+      }
+    } else if (data instanceof ArrowType.Utf8) {
+      return DataTypes.StringType;
+    } else if (data instanceof ArrowType.Binary) {
+      return DataTypes.BinaryType;
+    } else if (data instanceof ArrowType.Decimal) {
+      ArrowType.Decimal decimalData = (ArrowType.Decimal) data;
+      return new DecimalType(decimalData.getPrecision(), decimalData.getScale());
+    } else if (data instanceof ArrowType.Date && ((ArrowType.Date) data).getUnit() == DateUnit.DAY) {
+      return DataTypes.DateType;
+    } else if (data instanceof ArrowType.Timestamp && ((ArrowType.Timestamp) data).getUnit() == TimeUnit.MICROSECOND) {
+      return DataTypes.TimestampType;
+    }
+
+    throw new UnsupportedOperationException("Unsupported data type: " + data);
+  }
+
+  public DataType fromArrowField(Field field) {
```

Comment:
The data type methods could be static instead. Could you break this class into `SparkArrowTypeUtil` and something like `ArrowAllocation`?